MichiganDaily/cookie
# Allow for multiple html entry points

#### What's this PR do?

This PR allows multiple graphics to belong to one `cookie` repository — probably the biggest change to `cookie` in a while. 

This is accomplished by allowing several HTML files to be served through Parcel. The main change that drives this is [parcel serving](https://parceljs.org/features/cli/#entries) based on the `src/graphic/*.html` glob instead of `src/graphic/index.html`.

- Changes the config to have a new `entries` property that encapsulates several other properties (`hed`, `dek`, `byline`, `source`, `note`, and `illustrator_output_filename`.
- Changes the wrapper file to include a new entry point selector to toggle between different entries.
- Adds a new query parameter to set the current entry point such that refreshes will not always bounce to `index.html`.
- Upgrades `parcel-transformer-nunjucks` to allow for the passage of a filename variable to the nunjucks template, allowing for correct matching of the config to an HTML file.
- Removes `graphic.scss` import in `main.scss` such that `main.scss` is agnostic to entry point specific styling. The user may add this back but this seems like a reasonable default if different entry points will have slightly different styles.
   - The default `index.html` does include `graphic.scss` in the styles block.
- Moves image download code into a JS utility file to prevent code duplication across entry points.

#### Why are we doing this? How does it help us?

Prevents clutter in our organization when we bake several cookies. A story might have several related graphics so all of the related graphics should be stored in a central repository. These related graphics might have a lot of data or code overlap so it'd be useful to reduce code duplication as well.

#### How should this be manually tested?

1. Clone this repository and switch to this branch by running `git checkout -b multiple-entrypoints`.
2. First, test that things normally work with just one entry point. Try developing and building with `index.html`.
3. Test out all of the buttons to make sure things work.
4. Next, create a new HTML file in `src.graphic` and include this as the only content: 
    ```html
    {% extends "src/graphic/template/base.html" %}
    ```
5. Edit your config by adding a new property in `entries` with the key set to the HTML filename, and the value set to something like the `index.html` entry.
6. Restart your server and note that you can now choose an entry point. Toggling should update the document in the iframe. Click on the buttons to make sure things work for both entry points.
7. Try extending the base template by adding something in the new HTML file's `scripts` or `styles` block like so:
    ```html
    {% extends "src/graphic/template/base.html" %}

    {% block styles %}
    {{ super() }}
    <link rel="stylesheet" href="./css/test.scss" type="text/css" media="all" />
    {% endblock %}

    {% block scripts %}
    <script type="module" src="./js/test.js" charset="utf-8"></script>
    {% endblock %}
    ```
8. Modify those files to see if different styles and scripts apply to different entry points.

It might be worthwhile to test if something like this works in <https://github.com/MichiganDaily/primary-22-visuals>

#### Are there any smells or added technical debt to note?

1. When you add or delete a new HTML entry point, you'll probably have to restart the server for changes to properly be reflected.
10. You can add a new entry point in `config.json` without a corresponding HTML file and you'll be allowed to pick the non-existent entry point.

There are probably more smells that I haven't encountered yet.

#### What are relevant issues or links?

#33 and <https://github.com/MichiganDaily/parcel-transformer-nunjucks/pull/12>

#### Have you done the following, if applicable:


#### TODO

* [x] Merge <https://github.com/MichiganDaily/parcel-transformer-nunjucks/pull/12>
* [x] Bump `parcel-transformer-nunjucks` version
* [ ] Update version after merge
